### PR TITLE
use transaction api instead of form when creating transaction

### PR DESCRIPTION
**Pull request by atimos**

### CODE REVIEW

File: includes/class-wc-mondido-transaction.php [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+
+		$subscriptions = $this->get_subscriptions($order);
+
+		if (is_wp_error($subscriptions)) {
+			return $subscriptions;
+		}
+
+		$data = array_merge([
+			'amount' => $amount,
+			'vat_amount' => 0,
```
Comment (OliverMon) [inline includes/class-wc-mondido-transaction.php:37]: Ska denna alltid vara 0?

File: includes/class-wc-mondido-transaction.php [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+
+		$subscriptions = $this->get_subscriptions($order);
+
+		if (is_wp_error($subscriptions)) {
+			return $subscriptions;
+		}
+
+		return $this->api->update_transaction($transaction_id, array_merge([
+			'amount' => $amount,
+			'vat_amount' => 0,
```
Comment (OliverMon) [inline includes/class-wc-mondido-transaction.php:96]: Samma här

File: includes/class-wc-gateway-mondido-hw.php [hunk 3]
```diff
@@ -269,10 +257,56 @@ public function process_payment( $order_id ) {
...
+					$this->payment_action === 'authorize',
+					$this->secret,
+					$this->getCustomerReference($order)
+				);
+			}
+		} else {
+			$transaction = $this->transaction->create(
+				$order,
+				$this->merchant_id,
+				$this->testmode === 'yes',
```
Comment (OliverMon) [inline includes/class-wc-gateway-mondido-hw.php:280]: Always testmode?

File: includes/class-wc-gateway-mondido-hw.php [hunk 4]
```diff
@@ -269,10 +257,56 @@ public function process_payment( $order_id ) {
...
+
+		if ($transaction_id) {
+			$transaction = $this->transaction->get($transaction_id);
+
+			if (!is_wp_error($transaction)) {
+				$transaction = $this->transaction->update(
+					$transaction->id,
+					$order,
+					$this->merchant_id,
+					$this->testmode === 'yes',
```
Comment (OliverMon) [inline includes/class-wc-gateway-mondido-hw.php:270]: Always testmode?

File: includes/abstracts/class-wc-gateway-mondido-abstract.php [hunk 5]
```diff
@@ -318,17 +113,12 @@ public function lookupTransactionByOrderId( $order_id ) {
...
-				'amount' => number_format( $amount, 2, '.', '' )
-			)
-		) );
+		$result = $this->transaction->capture($transaction_id, $amount);
+		if (is_wp_error($result)) {
+			return $result;
+		}
 
-		return $result;
+		return json_decode(json_encode($result), true);
```
Comment (OliverMon) [inline includes/abstracts/class-wc-gateway-mondido-abstract.php:121]: Why?

File: includes/abstracts/class-wc-gateway-mondido-abstract.php [hunk 6]
```diff
@@ -226,87 +86,22 @@ public function add_order_fee($fee, &$order, $qty = 1) {
...
-			if (count($data) === 0) {
-				return null;
-			}
-			return $data[0];
+		$transaction = $this->transaction->get_by_reference($order_id);
+		if (is_wp_error($transaction)) {
+			wc_add_notice( $transaction->get_error_message(), 'error' );
+			return false;
 		}
+		return json_decode(json_encode($transaction), true);
```
Comment (OliverMon) [inline includes/abstracts/class-wc-gateway-mondido-abstract.php:104]: Why encode and decode?

File: includes/abstracts/class-wc-gateway-mondido-abstract.php [hunk 7]
```diff
@@ -226,87 +86,22 @@ public function add_order_fee($fee, &$order, $qty = 1) {
...
-
-			wc_add_notice( $e->getMessage(), 'error' );
+		$transaction = $this->transaction->get($transaction_id);
+		if (is_wp_error($transaction)) {
+			wc_add_notice( $transaction->get_error_message(), 'error' );
+			return false;
 		}
-
-		return FALSE;
+		return json_decode(json_encode($transaction), true);
```
Comment (OliverMon) [inline includes/abstracts/class-wc-gateway-mondido-abstract.php:95]: Same as other comments

File: includes/class-wc-mondido-transaction.php (hunk 1 above)
Comment (atimos) [reply includes/class-wc-mondido-transaction.php:38]: yes, det är samma som:
https://github.com/Mondido/MondidoWooCommerce/pull/62/files/4b4cc10e93b0c1ce0c6514c4d9f34ab1cf887bc6#diff-b54747cb336a9d197f36636168ad14bcL319

File: includes/class-wc-mondido-transaction.php (hunk 2 above)
Comment (atimos) [reply includes/class-wc-mondido-transaction.php:98]: samma svar ^^

File: includes/class-wc-gateway-mondido-hw.php (hunk 3 above)
Comment (atimos) [reply includes/class-wc-gateway-mondido-hw.php:282]: its only true if testmode has value of "yes"

File: includes/class-wc-gateway-mondido-hw.php (hunk 4 above)
Comment (atimos) [reply includes/class-wc-gateway-mondido-hw.php:271]: its only true if testmode has value of "yes"

File: includes/abstracts/class-wc-gateway-mondido-abstract.php (hunk 6 above)
Comment (aait) [reply includes/abstracts/class-wc-gateway-mondido-abstract.php:104]: to convert object to array ;-)

File: includes/abstracts/class-wc-gateway-mondido-abstract.php (hunk 5 above)
Comment (atimos) [reply includes/abstracts/class-wc-gateway-mondido-abstract.php:121]: the old api call decoded everything to an array, the new one is decoding to object, so to not break old code i just replaced the inner of the old functions with the new api and reencoded the result to array, to keep the old behaviour.
new code is using the transaction property directly

File: includes/abstracts/class-wc-gateway-mondido-abstract.php (hunk 6 above)
Comment (atimos) [reply includes/abstracts/class-wc-gateway-mondido-abstract.php:104]: the old api call decoded everything to an array, the new one is decoding to object, so to not break old code i just replaced the inner of the old functions with the new api and reencoded the result to array, to keep the old behaviour.
new code is using the transaction property directly

File: includes/abstracts/class-wc-gateway-mondido-abstract.php (hunk 7 above)
Comment (atimos) [reply includes/abstracts/class-wc-gateway-mondido-abstract.php:95]: same answer :)

File: includes/class-wc-mondido-transaction.php (hunk 1 above)
Comment (OliverMon) [reply includes/class-wc-mondido-transaction.php:38]: Sjukt

**Comment by atimos** [comment]
> What's the expected output if fee would be 100 and qty 3? Where would we cut the price and in what way etc?
> My past experience is that this might become 3x33 where you the lost 1 on the way. That in turn means you'll have to adjust one of the fee's up one (and in some occasions you'll have to lower it).
> 
> So have we tested this so that this isn't the case or would it be wise to implement a check that we're not losing/adding amount when doing this type of equations (and then dealing with rounding)

The fee amount is calculated in backend by multiplying quantity and plan_price, so 100 amount qty 3 is not very likely to happen.
We could get a weird price if the plan_price is unusual, but i dont think that is going to happen in practice.
And the total amount on the order is always going to be correct. So if it happens it is just going to be a visual error.

The fees in woo dont support quantity, but i could write the quantity on one fee line, something like "3 x Subscription #102325" if you are not sure about the current solution

**Comment by OliverMon** [comment]
If we know it wont happen - then there's no issue. Just so that we've covered all possibilities of messing up the total amount in Woo being something and the total amount that we try to charge the customer is something else.

**Review by OliverMon — APPROVED** [review]

**Comment by atimos** [comment]
No, we are not calculating the total order amount from the items, we are forcing it to be the same as the transaction amount

**Pull request merged**